subctl/issue-182-subctl-a-kubectl-plugin.md

```diff
@@ -0,0 +1,108 @@
+# Support krew way of installing `subctl`
+
+Related Issue:
+[Add subctl as kubectl plugin](https://github.com/submariner-io/enhancements/issues/182)
+
+## Summary
+
+Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is
+similar to tools like apt, dnf or brew. Today, over [200 kubectl plugins](https://krew.sigs.k8s.io/plugins/) are available on Krew.
+
+A plugin is a standalone executable file, whose name begins with `kubectl-`.
+
+More information can be found on [extending kubectl with plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) page.
+
+On the surface, installing a kubectl plugin seems simple enough – all you need to do is to place an executable in the user’s `PATH` prefixed
+with `kubectl-` – that you may be considering some other alternatives to Krew, such as:
+
+* Having the user manually download the plugin executable and move it to some directory in the PATH
+* Distributing the plugin executable using an OS package manager, like Homebrew (macOS), apt/yum (Linux), or Chocolatey (Windows)
+* Distributing the plugin executable using a language package manager (e.g. npm or go get)
+
+While these approaches are not necessarily unworkable, potential drawbacks to consider include:
+
+* How to get updates to users (in the case of manual installation)
+* How to package a plugin for multiple platforms (macOS, Linux, and Windows)
+* How to ensure your users have the appropriate language package manager (go, npm)
+* How to handle a change to the implementation language (e.g. a move from npm to another package manager)
+
+Krew solves these problems cleanly for all kubectl plugins, since it’s designed specifically to address these shortcomings. With Krew,
+after you write a plugin manifest once your plugin can be installed on all platforms without having to deal with their package managers.
+
+## Proposal
+
+A `kubectl-subm` binary would be created and its krew manifest file written.
+
+## Design
+
+To generate `kubectl-subm` binary, the current [`dist/subctl-%.tar.xz`](https://github.com/submariner-io/subctl/blob/devel/Makefile#L92)
+target in subctl repository will be edited to also create `kubectl-subm*.tar.gz` binaries and their checksum file. This is what
+[Krew expects](https://krew.sigs.k8s.io/docs/developer-guide/plugin-manifest/#sample-plugin-manifest).
+
+    dist/subctl-%: cmd/bin/subctl-%
+    mkdir -p dist
+    tar -cJf $@.tar.xz --transform "s/^cmd.bin/subctl-$(VERSION)/" $<
+    tar -cJf $@.tar.gz --transform "s/^cmd.bin.subctl-$(VERSION)/kubectl-subm-$(VERSION)/" $<
```

Comment:
This should use a different archive name, not just a different extension — the convention is that tarballs with varying compression types ship the same contents, whereas here we end up with a `.gz` tarball containing only the plugin.

Reply:
tar.xz and tar.gz both have `subctl` binareis. A plugin, in our case, is nothing but a binary named `kuectl-subm`. But it makes more sense to rename the `.gz` to `kubectl-subm.tar.gz` to differentiate the binary used as plugin.

Reply:
I know the desired result is the same binary with different names; however the convention is that `tar xf` with two tarballs with the same name but a different compression method should produce exactly the same result, names included, which wouldn’t be the case here. `kubectl-subm.tar.gz` is fine.

Note that the `tar` command given doesn’t work; perhaps in the proposal it would be best to describe the desired result, and leave the implementation details to the actual PR implementing the proposal.